From c4644e1f981dc572f33a9e4a43269f234551bccc Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Thu, 13 Oct 2022 09:26:16 +0800
Subject: [PATCH 49/49] lavc/qsvenc_av1: add missed options

QSV AV1 encoder may support -extbrc and -look_ahead_depth
---
 doc/encoders.texi       | 6 ++++++
 libavcodec/qsvenc.c     | 2 ++
 libavcodec/qsvenc_av1.c | 2 ++
 3 files changed, 10 insertions(+)

diff --git a/doc/encoders.texi b/doc/encoders.texi
index 4848e812f6..c999546e3f 100644
--- a/doc/encoders.texi
+++ b/doc/encoders.texi
@@ -3799,6 +3799,12 @@ This flag controls changing of frame type from B to P.
 
 @item @var{b_strategy}
 This option controls usage of B frames as reference.
+
+@item @var{extbrc}
+Extended bitrate control.
+
+@item @var{look_ahead_depth}
+Depth of look ahead in number frames, available when extbrc option is enabled.
 @end table
 
 @section snow
diff --git a/libavcodec/qsvenc.c b/libavcodec/qsvenc.c
index 09777414f2..6de4811028 100644
--- a/libavcodec/qsvenc.c
+++ b/libavcodec/qsvenc.c
@@ -1012,6 +1012,8 @@ static int init_video_param(AVCodecContext *avctx, QSVEncContext *q)
 
             q->extparam_internal[q->nb_extparam_internal++] = (mfxExtBuffer *)&q->extco2;
         } else if (avctx->codec_id == AV_CODEC_ID_AV1) {
+            if (q->extbrc >= 0)
+                q->extco2.ExtBRC = q->extbrc ? MFX_CODINGOPTION_ON : MFX_CODINGOPTION_OFF;
             if (q->b_strategy >= 0)
                 q->extco2.BRefType = q->b_strategy ? MFX_B_REF_PYRAMID : MFX_B_REF_OFF;
             if (q->adaptive_i >= 0)
diff --git a/libavcodec/qsvenc_av1.c b/libavcodec/qsvenc_av1.c
index 6abad91a19..5e6cfac0aa 100644
--- a/libavcodec/qsvenc_av1.c
+++ b/libavcodec/qsvenc_av1.c
@@ -109,11 +109,13 @@ static const AVOption options[] = {
     QSV_OPTION_B_STRATEGY
     QSV_OPTION_ADAPTIVE_I
     QSV_OPTION_ADAPTIVE_B
+    QSV_OPTION_EXTBRC
     { "profile", NULL, OFFSET(qsv.profile), AV_OPT_TYPE_INT, { .i64 = MFX_PROFILE_UNKNOWN }, 0, INT_MAX, VE, "profile" },
         { "unknown" , NULL, 0, AV_OPT_TYPE_CONST, { .i64 = MFX_PROFILE_UNKNOWN      }, INT_MIN, INT_MAX,     VE, "profile" },
         { "main"    , NULL, 0, AV_OPT_TYPE_CONST, { .i64 = MFX_PROFILE_AV1_MAIN     }, INT_MIN, INT_MAX,     VE, "profile" },
     { "tile_cols",  "Number of columns for tiled encoding",   OFFSET(qsv.tile_cols),    AV_OPT_TYPE_INT, { .i64 = 0 }, 0, UINT16_MAX, VE },
     { "tile_rows",  "Number of rows for tiled encoding",      OFFSET(qsv.tile_rows),    AV_OPT_TYPE_INT, { .i64 = 0 }, 0, UINT16_MAX, VE },
+    { "look_ahead_depth", "Depth of look ahead in number frames, available when extbrc option is enabled", OFFSET(qsv.look_ahead_depth), AV_OPT_TYPE_INT, { .i64 = 0 }, 0, 100, VE },
     { NULL },
 };
 
-- 
2.17.1

